osu.Game/Beatmaps/ControlPoints/ControlPoint.cs

```diff
@@ -5,13 +5,20 @@
 
 namespace osu.Game.Beatmaps.ControlPoints
 {
-    public class ControlPoint : IComparable<ControlPoint>
+    public class ControlPoint : IComparable<ControlPoint>, IEquatable<ControlPoint>
     {
         /// <summary>
         /// The time at which the control point takes effect.
         /// </summary>
         public double Time;
 
         public int CompareTo(ControlPoint other) => Time.CompareTo(other.Time);
+
+        public bool Equals(ControlPoint other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+            return Time.Equals(other.Time);

```

Comment:
Hmm okay,

However I think this is still going to break in the case of e.g. comparing an `EffectControlPoint` and a `DifficultyControlPoint`. Perhaps consider adding `if (!ReferenceEquals(GetType(), other.GetType()) return false;`?